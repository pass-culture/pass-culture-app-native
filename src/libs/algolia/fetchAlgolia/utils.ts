export const buildHitsPerPage = (hitsPerPage: number | null) =>
  hitsPerPage ? { hitsPerPage } : null
