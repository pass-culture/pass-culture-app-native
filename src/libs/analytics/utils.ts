import { NativeSyntheticEvent, NativeScrollEvent } from 'react-native'

type Props = NativeSyntheticEvent<NativeScrollEvent>['nativeEvent'] & { padding?: number }

export const isCloseToBottom = ({
  layoutMeasurement,
  contentOffset,
  contentSize,
  padding = 20,
}: Props) => layoutMeasurement.height + contentOffset.y >= contentSize.height - padding

export const isCloseToEndHorizontal = ({
  layoutMeasurement,
  contentOffset,
  contentSize,
  padding = 0,
}: Props) => layoutMeasurement.width + contentOffset.x >= contentSize.width - padding

// We don't send integers to firebase because they will be cast into int_value, float_value,
// or double_value in BigQuery depending on its value. To facilitate the work of the team,
// we just cast it to string.
export const prepareLogEventParams = (params: Record<string, unknown>) =>
  Object.keys(params).reduce((acc: Record<string, unknown>, key) => {
    acc[key] = typeof params[key] === 'number' ? (params[key] as number).toString() : params[key]
    return acc
  }, {})
