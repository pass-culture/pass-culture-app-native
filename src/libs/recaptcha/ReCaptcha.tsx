import React, { useState, useRef } from 'react'
import { Modal } from 'react-native'
import WebView, { WebViewMessageEvent } from 'react-native-webview'
import { ShouldStartLoadRequest } from 'react-native-webview/lib/WebViewTypes'
import styled from 'styled-components/native'

import { env } from 'libs/environment'

import { reCaptchaWebviewHTML } from './webviewHTML'

const ORIGIN_WHITELIST = ['*']

const WEBVIEW_SOURCE = {
  html: reCaptchaWebviewHTML,
  baseUrl: env.WEBAPP_URL,
}

type MessagePayload =
  | { message: 'close' }
  | { message: 'error'; error: string }
  | { message: 'expire' }
  | { message: 'load' }
  | { message: 'success'; token: string }

type Props = {
  onClose: () => void
  onError: (error: string) => void
  onExpire: () => void
  onLoad?: () => void
  onSuccess: (token: string) => void
  isVisible: boolean
}

export function ReCaptcha(props: Props) {
  const webViewRef = useRef<WebView>(null)
  const [isLoading, setIsLoading] = useState(true)

  function onLoadStart() {
    setIsLoading(true)
  }

  function handleLoad() {
    props.onLoad && props.onLoad()
    setIsLoading(false)
  }

  function handleMessage(event: WebViewMessageEvent) {
    try {
      const payload: MessagePayload = JSON.parse(event.nativeEvent.data)
      if (payload.message === 'load') handleLoad()
      if (payload.message === 'close') props.onClose()
      if (payload.message === 'expire') props.onExpire()
      if (payload.message === 'error') props.onError(payload.error)
      if (payload.message === 'success') props.onSuccess(payload.token)
    } catch (error) {
      console.warn(error)
    }
  }

  function handleNavigationStateChange() {
    // prevent navigation on Android
    if (!isLoading) {
      webViewRef.current?.stopLoading()
    }
  }

  function handleShouldStartLoadWithRequest(event: ShouldStartLoadRequest) {
    // prevent navigation on iOS
    return event.navigationType === 'other'
  }

  return (
    <Modal
      onRequestClose={props.onClose}
      testID="recaptcha-webview-modal"
      transparent
      visible={props.isVisible}>
      <StyledWebview
        allowsBackForwardNavigationGestures={false}
        bounces={false}
        onError={(event) => {
          const errorMessage = 'WebView error : ' + event.nativeEvent.description
          props.onError(errorMessage)
        }}
        onLoadStart={onLoadStart}
        onMessage={handleMessage}
        onNavigationStateChange={handleNavigationStateChange}
        onShouldStartLoadWithRequest={handleShouldStartLoadWithRequest}
        originWhitelist={ORIGIN_WHITELIST}
        ref={webViewRef}
        source={WEBVIEW_SOURCE}
        testID="recaptcha-webview"
      />
    </Modal>
  )
}

const StyledWebview = styled(WebView)({
  flex: 1,
  backgroundColor: 'rgba(0, 0, 0, 0)',
})
