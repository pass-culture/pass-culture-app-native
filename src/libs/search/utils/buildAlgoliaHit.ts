import { ResultItem } from '@elastic/app-search-javascript'

import { CategoryNameEnum } from 'api/gen'
import { AlgoliaHit } from 'libs/algolia'
import { SuggestedPlace } from 'libs/place'
import { AppSearchFields, TRUE } from 'libs/search/filters/constants'

// TODO(antoinewg) We need this function temporarily but delete when we migrate completely to App Search
export const buildAlgoliaHit = (searchHit: ResultItem<AppSearchFields>): AlgoliaHit => {
  const dates = (searchHit.getRaw(AppSearchFields.dates) as string[]).map(
    (ts: string) => new Date(ts).getTime() / 1000
  )
  const prices = (searchHit.getRaw(AppSearchFields.prices) as string[]).map((p: string) => +p / 100)
  const geoloc = searchHit.getRaw(AppSearchFields.venue_position) as string
  const [lat, lng] = (geoloc || ',').split(',')

  return {
    offer: {
      category: searchHit.getRaw(AppSearchFields.category) as CategoryNameEnum,
      dates,
      description: searchHit.getRaw(AppSearchFields.description) as string,
      isDigital: +searchHit.getRaw(AppSearchFields.is_digital) === TRUE,
      isDuo: +searchHit.getRaw(AppSearchFields.is_duo) === TRUE,
      name: searchHit.getRaw(AppSearchFields.name) as string,
      prices,
      thumbUrl: searchHit.getRaw(AppSearchFields.thumb_url) as string,
    },
    _geoloc: {
      lat: isNaN(parseFloat(lat)) ? null : parseFloat(lat),
      lng: isNaN(parseFloat(lng)) ? null : parseFloat(lng),
    },
    objectID: searchHit.getRaw(AppSearchFields.id) as string,
  }
}

export const buildVenues = (searchHit: ResultItem<AppSearchFields>): SuggestedPlace => {
  const geoloc = searchHit.getRaw(AppSearchFields.venue_position) as string
  const [lat, lng] = (geoloc || ',').split(',')
  const latitude = isNaN(parseFloat(lat)) ? null : parseFloat(lat)
  const longitude = isNaN(parseFloat(lng)) ? null : parseFloat(lng)

  const venueId = searchHit.getRaw(AppSearchFields.venue_id) as string

  return {
    label: searchHit.getRaw(AppSearchFields.venue_name) as string,
    info: searchHit.getRaw(AppSearchFields.offerer_name) as string,
    geolocation: latitude && longitude ? { latitude, longitude } : null,
    venueId: isNaN(parseFloat(venueId)) ? null : parseFloat(venueId),
  }
}
