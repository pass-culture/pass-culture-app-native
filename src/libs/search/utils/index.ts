export { filterAvailableCategories } from './filterAvailableCategories'
export { getCategoriesFacetFilters } from './getCategoriesFacetFilters'
