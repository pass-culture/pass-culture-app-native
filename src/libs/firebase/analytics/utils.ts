import { NativeSyntheticEvent, NativeScrollEvent } from 'react-native'

import { SearchState } from 'features/search/types'
import { PerformSearchState } from 'libs/firebase/analytics/types'

type Props = NativeSyntheticEvent<NativeScrollEvent>['nativeEvent'] & { padding?: number }

export const isCloseToBottom = ({
  layoutMeasurement,
  contentOffset,
  contentSize,
  padding = 20,
}: Props) => layoutMeasurement.height + contentOffset.y >= contentSize.height - padding

export const isCloseToEndHorizontal = ({
  layoutMeasurement,
  contentOffset,
  contentSize,
  padding = 0,
}: Props) => layoutMeasurement.width + contentOffset.x >= contentSize.width - padding

// We don't send integers to firebase because they will be cast into int_value, float_value,
// or double_value in BigQuery depending on its value. To facilitate the work of the team,
// we just cast it to string.
export const prepareLogEventParams = (params: Record<string, unknown>) =>
  Object.keys(params).reduce((acc: Record<string, unknown>, key) => {
    acc[key] = typeof params[key] === 'number' ? (params[key] as number).toString() : params[key]
    return acc
  }, {})

const STRING_VALUE_MAX_LENGTH = 100
export const urlWithValueMaxLength = (url: string) => url.slice(0, STRING_VALUE_MAX_LENGTH)

export const buildPerformSearchState = (searchState: SearchState) => {
  const state: PerformSearchState = {
    searchLocationFilter: JSON.stringify(searchState.locationFilter),
    searchId: searchState.searchId,
  }

  if (searchState.date !== null) {
    state.searchDate = JSON.stringify(searchState.date)
  }

  if (searchState.isAutocomplete) {
    state.searchIsAutocomplete = searchState.isAutocomplete
  }

  if (searchState.maxPrice) {
    state.searchMaxPrice = searchState.maxPrice
  }

  if (searchState.minPrice) {
    state.searchMinPrice = searchState.minPrice
  }

  if (searchState.offerCategories.length > 0) {
    state.searchCategories = JSON.stringify(searchState.offerCategories)
  }

  if (searchState.offerGenreTypes && searchState.offerGenreTypes.length > 0) {
    state.searchGenreTypes = JSON.stringify(searchState.offerGenreTypes)
  }

  if (searchState.offerIsDuo) {
    state.searchOfferIsDuo = searchState.offerIsDuo
  }

  if (searchState.offerIsFree) {
    state.searchOfferIsFree = searchState.offerIsFree
  }

  if (searchState.offerNativeCategories && searchState.offerNativeCategories.length > 0) {
    state.searchNativeCategories = JSON.stringify(searchState.offerNativeCategories)
  }

  if (searchState.query !== '') {
    state.searchQuery = searchState.query
  }

  if (searchState.timeRange !== null) {
    state.searchTimeRange = JSON.stringify(searchState.timeRange)
  }

  return state
}
