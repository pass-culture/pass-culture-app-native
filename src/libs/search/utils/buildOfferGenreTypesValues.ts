import { OfferGenreType } from 'features/search/types'

export const buildOfferGenreTypesValues = ({
  movieGenres,
  musicTypes,
  showTypes,
}: {
  movieGenres: OfferGenreType[] | undefined
  musicTypes: OfferGenreType[] | undefined
  showTypes: OfferGenreType[] | undefined
}): OfferGenreType[] => {
  let finalOfferGenreType: OfferGenreType[] = []

  finalOfferGenreType = finalOfferGenreType?.concat(movieGenres ?? [])
  finalOfferGenreType = finalOfferGenreType?.concat(musicTypes ?? [])
  finalOfferGenreType = finalOfferGenreType?.concat(showTypes ?? [])

  return finalOfferGenreType
}
