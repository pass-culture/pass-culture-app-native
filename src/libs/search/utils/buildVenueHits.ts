import { ResultItem } from '@elastic/app-search-javascript'

import { VenueHit } from 'libs/search'
import { AppSearchVenuesFields } from 'libs/search/filters/constants'

export const buildVenueHits = (
  searchHit: ResultItem<AppSearchVenuesFields>
): Pick<VenueHit, 'id' | 'name' | 'venueType'> => {
  return {
    id: searchHit.getRaw(AppSearchVenuesFields.id) as string,
    name: searchHit.getRaw(AppSearchVenuesFields.name) as string,
    venueType: searchHit.getRaw(AppSearchVenuesFields.venue_type) as string,
  }
}
