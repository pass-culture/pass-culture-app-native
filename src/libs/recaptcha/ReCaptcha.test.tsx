import { render } from '@testing-library/react-native'
import React from 'react'

import { simulateWebviewMessage } from 'tests/utils'

import { ReCaptcha } from './ReCaptcha'

afterEach(jest.clearAllMocks)

const reCaptchaProps = {
  onClose: jest.fn(),
  onError: jest.fn(),
  onExpire: jest.fn(),
  onLoad: jest.fn(),
  onSuccess: jest.fn(),
  isVisible: true,
}

describe('<ReCaptcha />', () => {
  it('should render correctly', () => {
    const renderAPI = render(<ReCaptcha {...reCaptchaProps} />)
    expect(renderAPI).toMatchSnapshot()
  })

  it('should not render webview when modal is not visible', () => {
    const renderAPI = render(<ReCaptcha {...reCaptchaProps} isVisible={false} />)
    const recaptchaWebview = renderAPI.queryByTestId('recaptcha-webview')
    expect(recaptchaWebview).toBeFalsy()
  })

  it("should call onSuccess() callback when webview's message is success", () => {
    const renderAPI = render(<ReCaptcha {...reCaptchaProps} />)
    const recaptchaWebview = renderAPI.getByTestId('recaptcha-webview')

    simulateWebviewMessage(recaptchaWebview, '{ "message": "success", "token": "fakeToken" }')

    expect(reCaptchaProps.onSuccess).toBeCalledWith('fakeToken')
  })

  it("should call onError() callback when webview's message is error", () => {
    const renderAPI = render(<ReCaptcha {...reCaptchaProps} />)
    const recaptchaWebview = renderAPI.getByTestId('recaptcha-webview')

    simulateWebviewMessage(recaptchaWebview, '{ "message": "error", "error": "someError" }')

    expect(reCaptchaProps.onError).toBeCalledWith('someError')
  })

  it("should call onClose() callback when webview's message is close", () => {
    const renderAPI = render(<ReCaptcha {...reCaptchaProps} />)
    const recaptchaWebview = renderAPI.getByTestId('recaptcha-webview')

    simulateWebviewMessage(recaptchaWebview, '{ "message": "close" }')

    expect(reCaptchaProps.onClose).toBeCalled()
  })

  it("should call onExpire() callback when webview's message is expire", () => {
    const renderAPI = render(<ReCaptcha {...reCaptchaProps} />)
    const recaptchaWebview = renderAPI.getByTestId('recaptcha-webview')

    simulateWebviewMessage(recaptchaWebview, '{ "message": "expire" }')

    expect(reCaptchaProps.onExpire).toBeCalled()
  })

  it("should call onLoad() callback when webview's message is load", () => {
    const renderAPI = render(<ReCaptcha {...reCaptchaProps} />)
    const recaptchaWebview = renderAPI.getByTestId('recaptcha-webview')

    simulateWebviewMessage(recaptchaWebview, '{ "message": "load" }')

    expect(reCaptchaProps.onLoad).toBeCalled()
  })
})
