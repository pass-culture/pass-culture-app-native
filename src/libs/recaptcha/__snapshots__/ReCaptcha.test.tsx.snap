// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReCaptcha /> should render correctly 1`] = `
<Modal
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  testID="recaptcha-webview-modal"
  transparent={true}
  visible={true}
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <RNCWebView
      allowsBackForwardNavigationGestures={false}
      bounces={false}
      cacheEnabled={true}
      injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
      injectedJavaScriptForMainFrameOnly={true}
      javaScriptEnabled={true}
      messagingEnabled={true}
      onContentProcessDidTerminate={[Function]}
      onError={[Function]}
      onHttpError={[Function]}
      onLoadStart={[Function]}
      onLoadingError={[Function]}
      onLoadingFinish={[Function]}
      onLoadingProgress={[Function]}
      onLoadingStart={[Function]}
      onMessage={[Function]}
      onNavigationStateChange={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      source={
        Object {
          "baseUrl": undefined,
          "html": "
    <!DOCTYPE html>
    <html lang=\\"fr\\">
    
    <head>
        <meta charset=\\"UTF-8\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
        <title></title>
        <script src=\\"https://www.google.com/recaptcha/api.js?hl={{lang}}\\" async defer></script>
        <script>
            let numberOfRetryRender = 0;
            let readyInterval;
            let onCloseInterval;
            let widget;
            let onCloseObserver;
            
            function sendMessagePayload(payload) {
                window.ReactNativeWebView.postMessage(JSON.stringify(payload));
            }
            function onClose() {
                sendMessagePayload({ \\"message\\": \\"close\\" });
            }
            function onLoad() {
                sendMessagePayload({ \\"message\\": \\"load\\" });
            }
            function onExpire() {
                sendMessagePayload({ \\"message\\": \\"expire\\" });
            }
            function onError(error) {
                sendMessagePayload({ \\"message\\": \\"error\\", error: error, });
            }
            function onSuccess(token) {
                sendMessagePayload({ \\"message\\": \\"success\\",  token: token, });
            }
    
            function isReady() { 
                return Boolean(typeof window === \\"object\\" && window.grecaptcha && window.grecaptcha.render);
            }
    
            function registerOnCloseListener() {
                if (onCloseObserver) {
                    onCloseObserver.disconnect();
                }
    
                const iframes = document.getElementsByTagName(\\"iframe\\");
    
                const recaptchaFrame = Array.prototype.find
                    .call(iframes, e => e.src.includes(\\"google.com/recaptcha/api2/bframe\\"));
                const recaptchaElement = recaptchaFrame.parentNode.parentNode;
    
                clearInterval(onCloseInterval);
    
                let lastOpacity = recaptchaElement.style.opacity;
                onCloseObserver = new MutationObserver(mutations => {
                    if (lastOpacity !== recaptchaElement.style.opacity
                        && recaptchaElement.style.opacity == 0) {
                        onClose();
                    }
                    lastOpacity = recaptchaElement.style.opacity;
                });
                onCloseObserver.observe(recaptchaElement, {
                    attributes: true,
                    attributeFilter: [\\"style\\"],
                });
            }
    
            function renderRecaptcha() {
                const options = {
                    sitekey: \\"undefined\\",
                    size: \\"invisible\\",
                    theme: \\"light\\",
                    callback: onSuccess,
                    \\"expired-callback\\": onExpire,
                    \\"error-callback\\": onError,
                }
                widget = window.grecaptcha.render(\\"recaptcha-container\\", options);
                window.grecaptcha.execute(widget);
                if (onLoad) {
                    onLoad();
                }
                onCloseInterval = setInterval(registerOnCloseListener, 1000);
            }

            function updateReadyState() {
                numberOfRetryRender = numberOfRetryRender + 1;
                if (isReady()) {
                    clearInterval(readyInterval);
                    renderRecaptcha();
                } else if (numberOfRetryRender === 6) {
                    clearInterval(readyInterval);
                    onError(\\"Number of retry render captcha exceeded\\");
                }
            }
        
            try {
                if (isReady()) {
                    renderRecaptcha();
                } else {
                    readyInterval = setInterval(updateReadyState, 1000);
                }
            } catch(error) {
                onError(error.message);
            }
        </script>
    
        <style>
            html,
            body,
            .container {
                height: 100%;
                width: 100%;
                margin: 0;
                padding: 0;
                background-color: transparent;
            }
    
            .container {
                display: flex;
                justify-content: center;
                align-items: center;
            }
        </style>
    </head>
    
    <body>
        <div class=\\"container\\">
            <span id=\\"recaptcha-container\\"></span>
        </div>
    </body>
    
    </html>",
        }
      }
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#ffffff",
          },
          Array [
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0)",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ],
        ]
      }
      testID="recaptcha-webview"
      useSharedProcessPool={true}
    />
  </View>
</Modal>
`;
